### ISSUE TITLE
Output of the pre-trained expert discriminator

### ISSUE THREAD

**Issue by Alpe6825** [Rudrabha/Wav2Lip#181]
I’ve a problem regarding the return value of `get_sync_loss(mel, g)` in `hq_wav2lip_train.py`.

If I understood the paper correctly, the pre-trained discriminator should decide if a given mel spectrogram and a sequence of five frames (lower face) are in sync or not, right? 

I want to test this independently and compared three different inputs:
1.	mel and synced sequence
2.	mel and out-of-sync sequence
3.	mel and sequence generated by an untrained generator 

I used a batch size of 2.

![Folie1](https://user-images.githubusercontent.com/51789043/103317813-a5772f00-4a2c-11eb-87bb-314c3a13df3e.jpeg)

I tried ten of this batches and in most cases the discriminator chose the frames which are out of sync. But I’m also wondering about the distribution of all returning values. I expected that good values are in a near of zero, in order to reduce the loss value. But that noisy sequences leads to a lower loss than synced sequences seems strange.

| synced | out-of-sync | untrained generator |
|---|---|---|
| **0.36840230226516724** | 2.3156867027282715 | 1.6549360752105713 |
|1.0246469974517822 |**0.2813396155834198** |50.29334259033203|
|0.9422204494476318 |1.4812026023864746 |**0.2827274799346924**|
|0.7599254846572876 |**0.40085890889167786**| 0.6270509958267212|
|0.9634140729904175 |**0.25472718477249146**| 0.8915684223175049|
|2.4803733825683594| **0.861980140209198**| 2.780829906463623|
|1.0815658569335938 |1.0694379806518555 |**0.582244873046875**|
|0.6146743893623352| **0.5238527655601501**| 0.61577308177948|
|0.5758190155029297 |1.371729850769043 |**0.45662859082221985**|
|1.474782109260559| 1.2921059131622314| **1.1070760488510132**|

Does anyone have any idea what the problem could be?
Or could anyone produce some return values for synced and out-of-sync sequences with the LRS2 dataset? It would be a great help.

**Comment by prajwalkr** [comment]
> But that noisy sequences leads to a lower loss than synced sequences seems strange.

The discriminator is trained only on real faces. As the generator starts generating more and more realistic faces, the discriminator becomes better at judging it. That is one of the reasons why the sync loss weight is 0 initially, and is increased after the generator is trained to an extent.

**Comment by Alpe6825** [comment]
> The discriminator is trained only on real faces. As the generator starts generating more and more realistic faces, the discriminator becomes better at judging it. That is one of the reasons why the sync loss weight is 0 initially, and is increased after the generator is trained to an extent.

The training process without the sync loss works well. The reason why I tested the expert discriminator seperatly is that with `syncnet_wt = 0` the training process is stable, but when the syncnet_wt gets greater 0 after some epochs the process tips over and the results looks like the following, even they looked realistic the epochs before without sync loss.

<img width="369" alt="Bildschirmfoto 2020-12-30 um 11 35 41" src="https://user-images.githubusercontent.com/51789043/103346504-7399b300-4a94-11eb-96d6-dbc0b4d1f9b7.png">

When the generator produces realistic faces they should be similar to the synced one. But when I tested the expert discriminator separately, it does not detect the synced one in a right way. Worser, it abuse them, cause only at the first row the synced one have the lowest loss. I think that's the reason for bad results. The question is, why doesn't the expert discriminator detect the synced sequences correctly?

**Comment by prajwalkr** [comment]
What dataset are you training on?

**Comment by Alpe6825** [comment]
Since I currently have no access to the LRS2 dataset, I used a small selfmade dataset of one person.

A person looks frontal into the camera and spokes some [Harvard Sentences](https://www.cs.columbia.edu/~hgs/audio/harvard.html). The video was recorded with 30fps but I converted it into 25fps. And I cropped the video to a square around the face.

In the original code from you, sync loss is used when `average_sync_loss <  .75`. This is never happens in my case. I changed that little bit. So the sync_loss is used after some epochs when the generated images looks realistic. 

`average_sync_loss` is constantly greater 1, if used the original synced frame sequences. According to that `average_sync_loss` for generated images are greater 1 too.

**Comment by prajwalkr** [comment]
We do not know if the model will work for limited data of a single speaker. Nevertheless, you need to re-train the SyncNet model before using it for Wav2Lip.

**Comment by cagrigungor** [comment]
Sir, first of all, I would like to congratulate you on this amazing research. I am planning to use your pre-trained lipsync_expert for different research. But, while I am testing, pre-trained lipsync_expert did not give expected losses. For example, I feeding sync net with sync and out of sync samples, but I get similar losses which are not expected. For my research, I am planning to use a different dataset called Crema-D, should I train syncnet with Crema-D to get meaningful results? I supposed that pretrained syncnet will work for the Crema-D dataset, but it does not work or I could not work. What is your opinion about my issue? Thank you.

**Comment by Alpe6825** [comment]
>  Nevertheless, you need to re-train the SyncNet model before using it for Wav2Lip.

Until now, I assumed that the trained SyncNet is generalized enough for training the other networks with only one person. 
I will try retraining the syncnet in the next days. Thx for the tipp.

**Comment by prajwalkr** [comment]
> For my research, I am planning to use a different dataset called Crema-D, should I train syncnet with Crema-D to get meaningful results?

Yes, you need to re-train it.

**Comment by zhanghming** [comment]
@Alpe6825 , Hi,

I also want to evaluate the performance of the LipSync Expert in my own dataset, but I have no idea how to compute the evaluation metric like yours, could you release your code in this issue to get your table results? 

BTW, are the values in your table the sync loss value? 

Thanks in advance.
